### PR TITLE
HBHE: New time slew parameters and upgrade of time slew class

### PR DESCRIPTION
**Pull request by cmadrid1**
This PR adds new parameters for the HBHE time slew delay function used for M2 and simulation. It also makes time slew an event producer. 

Validation:
https://indico.cern.ch/event/649519/contributions/2641903/attachments/1495366/2326436/TimeSlewCorrections.pdf

Minor improvements for HPD simulation and data digi comparison as well as lower chi2 for simulation reco. 

@hatakeyamak @kpedro88 @mariadalfonso

### CODE REVIEW

**Comment by kpedro88** [comment]
please test

**Comment by cmadrid1** [comment]
Ok the "HcalPulseContainmentTest" fails here because “HcalPulseContainmentManager” class needs to get the time slew object from the event setup before it is used:
https://github.com/cms-sw/cmssw/blob/09c3fce6626f70fd04223e7dacebf0b485f73f54/CalibCalorimetry/HcalAlgos/test/HcalPulseContainmentTest.cpp#L8

**Comment by abdoulline** [comment]
@cmadrid1  @kpedro88 
 I think HcalPulseCiontainmentManager needs some simple protection against unavailability of the HcalTimeSlewRecord
https://github.com/cms-sw/cmssw/pull/21733/files#diff-52f9a6773eadbbfb02cec893461488a4R18

If I remember it right, in the example below this kind of "workaround" for absent HcalMCParams conditions: 
http://cmslxr.fnal.gov/source/SimCalorimetry/HcalSimAlgos/src/HcalShapes.cc#0060
http://cmslxr.fnal.gov/source/SimCalorimetry/HcalSimAlgos/src/HcalShapes.cc#0084
had been put in place to avoid crashing corresponding AddOn test

**Comment by abdoulline** [comment]
@slava77, @perrotta  
Looks like it might now make it to pre3, 
even though apart from the failed AddOn test, the rest could hopefully be revised.

**Comment by slava77** [comment]
>  Looks like it might not make it to pre3,

@abdoulline 
please clarify if it is required for pre3.

**Comment by deguio** [comment]
Hi @slava77 
in principle yes pre3 was the original target. but I let @cmadrid1 and @hatakeyamak comment more on the status

**Comment by davidlange6** [comment]
I’m not sure how things are required for a prerelease, but building it after Thursday means effectively January 8


On 18 Dec 2017, at 18:34, Slava Krutelyov <notifications@github.com<mailto:notifications@github.com>> wrote:


Looks like it might not make it to pre3,

@abdoulline<https://github.com/abdoulline>
please clarify if it is required for pre3.

**Comment by abdoulline** [comment]
Sorry for coming late back to the thread... 
 We did plan to have all the DIGI-related 2018 updates in pre3, as Federico mentioned.

**Comment by kpedro88** [comment]
please test

**Comment by Martin-Grunewald** [comment]
-1

**Comment by Martin-Grunewald** [comment]
And please use fillDescriptions for new parameters so as to avoid customize!

**Comment by ghellwig** [comment]
@cmadrid1 unrelated to this PR, but since you are touching code in this file:
```
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-12-18-2300/src/RecoLocalCalo/HcalRecAlgos/src/HcalSimpleRecAlgo.cc:19:15: warning: unused variable 'HPDShapev3DataNum' [-Wunused-const-variable]
constexpr int HPDShapev3DataNum = 105;
              ^
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_10_0_X_2017-12-18-2300/src/RecoLocalCalo/HcalRecAlgos/src/HcalSimpleRecAlgo.cc:20:15: warning: unused variable 'HPDShapev3MCNum' [-Wunused-const-variable]
constexpr int HPDShapev3MCNum = 105;
              ^
2 warnings generated.
```

**Comment by kpedro88** [comment]
@Martin-Grunewald the customize is needed because configs in `HLTrigger/Configuration/test` don't pick up `Hcal_Conditions_forGlobalTag_cff`, which is where we put the `ESProducer` for `HcalTimeSlew`. If you have a better solution, please let us know.

**Comment by Martin-Grunewald** [comment]
It is wrong to do load the general GT module as done above.
Load the file with the actual specific ES module

**Comment by perrotta** [comment]
please test

**Comment by kpedro88** [comment]
@Martin-Grunewald it looks like you recently changed `customizeHLTforCMSSW` so the customizations only get called for specific menu types:
https://github.com/cms-sw/cmssw/blame/8c27d724e7babee7597b5fd34172c406e33e9e84/HLTrigger/Configuration/python/customizeHLTforCMSSW.py#L74

This means that some of the HLT test configs are not picking up our customization. What should we do about this?

**Comment by Martin-Grunewald** [comment]
Put the call after/outside the if-block, as also done in #21664 .
Of course these changes will step on each other so one PR needs to be fixed after the other is integrated.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by Martin-Grunewald** [comment]
+1

**Comment by Martin-Grunewald** [comment]
This PR has merge conflicts!

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
Uhm, 132 commits...
Maybe a rebase is needed?

**Comment by abdoulline** [comment]
@jaylawhorn, @cmadrid1  
I have the impression two branches are "glued" ?

**Comment by cmadrid1** [comment]
I am fixing it now.

**Comment by perrotta** [comment]
please test

File: RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc [hunk 1]
```diff
@@ -269,11 +271,11 @@ double MahiFit::minimize() const {
...
 
 void MahiFit::updatePulseShape(double itQ, FullSampleVector &pulseShape, FullSampleVector &pulseDeriv,
-			       FullSampleMatrix &pulseCov) const {
+			       FullSampleMatrix &pulseCov,
+			       const HcalTimeSlew* hcalTimeSlew_delay) const {
   
   float t0=meanTime_;
-  if (applyTimeSlew_) 
-    t0+=HcalTimeSlew::delay(std::max(1.0, itQ), slewFlavor_);
+  if (applyTimeSlew_) t0+=hcalTimeSlew_delay->delay(std::max(1.0, itQ), slewFlavor_);
```
Comment (jaylawhorn) [inline RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc]: Can we please retain the previous behavior? Possibly by adding a method like setPulseShapeTemplate [here](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc#L65)?

Comment (jaylawhorn) [reply RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc]: As long as this works I can mess with the technical parts later.

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

**Comment by cmadrid1** [comment]
That didn't work. I'm just going to rebase to pre3.

**Comment by perrotta** [comment]
Why do you want to rebase to pre3?
The conflict is with the HEAD of the master, i.e. CMSSW_10_0_X_2018-01-03-1100 + #21799

**Comment by cmadrid1** [comment]
Right. I will do that.

**Comment by perrotta** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by hatakeyamak** [comment]
I was wondering if this PR is being reviewed. We would like to have this merged in pre4. For 2017 and earlier scenarios, this PR shouldn't make any significant differences, and for 2018 and beyond, it should adopt the improved time slew parametrization. Also, the the way how time slew is adopted should be improved in this PR by using the EventSetup.

This PR will affect both digi and reco, as time slew class is used for both simulation and reconstruction steps.

This is a relatively large PR, so I wanted to make sure this won't be left until the last minute.

**Comment by perrotta** [comment]
Hi @hatakeyamak : the PR is being reviewed right now.

By the way: maybe the reason will be clearer to me when I'll dig deeper in the code, but from the outputs I see that there are changes even in the 2011 MinBias wf, see e.g.:
image

This apparently contradicts what you says about "2017 and earlier scenarios": could you please check and comment? Thank you.

**Comment by hatakeyamak** [comment]
Thank you for working on this review.

Hmm. Interesting. Do you see this difference only for eaux, but not for energy and eraw? Do you see a similar level of differences in some MC samples of 2017 or earlier?

**Comment by perrotta** [comment]
@hatakeyamak : you can see the differences yourself by looking at the results of the jenkins tests, as linked out in https://github.com/cms-sw/cmssw/pull/21733#issuecomment-355190420

The reco changes are in the following workflows:
```
/build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison
JR results differ 3 all_OldVSNew_Higgs200ChargedTauswf9p0
JR results differ 147 all_mini_OldVSNew_TTbar13TeV2018wf10824p0
JR results differ 5 all_OldVSNew_TTbarPUwf25202p0
JR results differ 275 all_OldVSNew_TTbar13TeV2019wf11624p0
JR results differ 7 all_OldVSNew_RunMinBias2011Awf1001p0
JR results differ 7 all_OldVSNew_RunMinBias2011Awf1000p0
JR results differ 138 all_mini_OldVSNew_TTbar13TeV2019wf11624p0
JR results differ 6 all_OldVSNew_TTbar13TeV2017wf10024p0
JR results differ 3 all_OldVSNew_TTbar13TeVPU2017wf10224p0
JR results differ 367 all_OldVSNew_TTbar13TeV2018wf10824p0
SUMMARY Reco comparison results: 958 differences found in the comparisons
```
where the corresponding plots with the diffs can be accessed from 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_0_X_2018-01-03-1100+21733/24458/validateJR/

The largest changes are in the 2018 and 2019 wfs, as expected. But still some regression is visible in 2017 and even 2011 ones (as the one I pointed out above).

As far as I can see, differences in <= 2017 workflows are only for eaux, and not for energy and eraw.

**Comment by hatakeyamak** [comment]
Okay. Thank you for this information. This probably means that we should look closely at the timeslew used for method3. We will keep you posted.

File: RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc [hunk 2]
```diff
@@ -121,35 +114,18 @@ void HcalDeterministicFit::phase1Apply(const HBHEChannelInfo& channelData,
...
-  }
-
-  if (fTimeSlew==0)respCorr=1.0;
-  else if (fTimeSlew==1) channelData.hasTimeInfo()?respCorr=rCorrSiPM[0]:respCorr=rCorr[0];
-  else if (fTimeSlew==2) channelData.hasTimeInfo()?respCorr=rCorrSiPM[1]:respCorr=rCorr[1];
-  else if (fTimeSlew==3)respCorr=frespCorr;
+  HcalTimeSlew::ParaSource subDet=HcalTimeSlew::ParaSource::HE;
+  if(std::abs(cell.ieta())<HcalRegion[0]) subDet=HcalTimeSlew::ParaSource::HB;
+  else if(std::abs(cell.ieta())==HcalRegion[0]||std::abs(cell.ieta())==HcalRegion[1]) subDet=HcalTimeSlew::ParaSource::BE;
+  else subDet=HcalTimeSlew::ParaSource::HE;
```
Comment (perrotta) [inline RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc]: This line is not needed

Comment (cmadrid1) [reply RecoLocalCalo/HcalRecAlgos/src/HcalDeterministicFit.cc]: I thought it wasn't needed but it was setup before to give different time slew parameters for HB and HE separately (However, they gave the same numbers). 
I will remove this from here and from the HcalTimeSlew delay function.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by perrotta** [comment]
please test

**Comment by cmadrid1** [comment]
Ok looks like the PR now only changes 2018 and beyond.

**Comment by perrotta** [comment]
I have tried to evaluate separately the changes done in step2 (i.e. simulation) and in step3 (i.e. reco). Therefore, I run the step3 of wf 10824 (TTbar 2018) with this PR on top of some step2 output (100 events) generated with the baseline ID; in parallel, I run the whole wf 10824 with this PR, i.e. sim-digi + reco.

In both cases, with 100 events a similar smear of the reco-related objects is visible, with no exceedingly significant change of shape  in any of those variables. No side effects on related quantities (as btag, lepton id and isolation, MET, etc.). With 100 events I would call it an "adjustment", while I expect that its full effect will be better visible in the validation, as it was already shown in the description part of this PR.

If anybody is interested, I've copied the differences in reco in
https://awagner.web.cern.ch/perrotta/TestPR21733_STEP3ONLY/all_testPR21733VSorig_TTbar13TeV2018wf10824p0
and
https://awagner.web.cern.ch/perrotta/TestPR21733_FULL/all_testPR21733VSorig_TTbar13TeV2018wf10824p0

**Comment by perrotta** [comment]
+1
- The adjustment to the HBHE time slew as implemented in this PR apparently propagates as expected into the HCAL local reco and down to all reco related quantities
- No negative side effects are visible, while the full evaluation requires some deeper validation as can be done by DPG and POGs
- After the last fixes, there is no effect on <= 2017 workflows, as it was in the original plans

**Comment by mariadalfonso** [comment]
@civanch, @ghellwig, @silviodonato, @arunhep, @Martin-Grunewald, @mdhildreth, @ggovi, @fwyzard, @lpernie
can you please have a look at this PR and eventually sign ?
we would like to get this PR into the pre4

**Comment by civanch** [comment]
+1

**Comment by arunhep** [comment]
+1

**Comment by ggovi** [comment]
+1

**Comment by cmadrid1** [comment]
Just rebased to fix a conflict with an HLT PR.

**Comment by Martin-Grunewald** [comment]
please test

**Comment by Martin-Grunewald** [comment]
+1

**Comment by Martin-Grunewald** [comment]
@fabiocos 
The four PRs #21845 #21842 #21821 #21733 all modify the file
HLTrigger/Configuration/python/customizeHLTforCMSSW.py.  Hence,
once one PR is merged, the others will show a merge conflict, and
so on.  The order of PR integration could be agreed beforehand
and the PRs rebased accordingly, or a single big PR could be
made.

**Comment by mariadalfonso** [comment]
@fabiocos the proposed order is the following  for merging is the following:
#21733 (time slew) READY change the local reconstruction and would be good to have merged , was signed already by everyone at least once
#21842 (CaloTower) still miss a commit, NEED A FIX was mentioned yesterday
#21845 (PFcluster) NEED A FIX, was mentioned yesterday
#21821 is pretty minor mainly code cleanup can go last
@deguio

**Comment by perrotta** [comment]
+1
- Only rebased since https://github.com/cms-sw/cmssw/pull/21733#issuecomment-357652165

**Comment by perrotta** [comment]
Wait a second!
I removed the "+1" because I want to understand why now all Phase2 workflows are also affected, which was not the case before the rebase: any quick explanation?

**Comment by perrotta** [comment]
+1
- Only rebased since https://github.com/cms-sw/cmssw/pull/21733#issuecomment-357652165
- Now also Phase2 workflows are affected because of #21804 (merged since CMSSW_10_0_X_2018-01-12-2300 ): starting from that IB the Phase2 era is based on run3 era, which is in turn based on run2_2018. Therefore, from that PR on, all modifications to the 2018 software will directly propagate into Phase2.

**Comment by kpedro88** [comment]
@perrotta to confirm, that was my intention

**Comment by lpernie** [comment]
+1

**Comment by fabiocos** [comment]
@ggovi @civanch can you please sign, unless you have comments/objections?

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
@ggovi please makes asap you remarks, or sign this PR, we need to close this development

**Comment by ggovi** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**